src/base/cub_ddl_log.c

```diff
@@ -0,0 +1,1206 @@
+	{
+	  snprintf (ddl_audit_handle->elapsed_time, sizeof (ddl_audit_handle->elapsed_time), "elapsed time 0.000");
+	}
+
+      retval = snprintf (msg, DDL_LOG_BUFFER_SIZE, "%s %s|%s|%s|%s|%s|%s\n",
+			 ddl_audit_handle->execute_start_time,
+			 ddl_audit_handle->ip_addr,
+			 ddl_audit_handle->user_name,
+			 result, ddl_audit_handle->elapsed_time, ddl_audit_handle->msg, ddl_audit_handle->sql_text);
+
+    }
+
+  if (retval >= DDL_LOG_BUFFER_SIZE)
+    {
+      msg[DDL_LOG_BUFFER_SIZE - 2] = '\n';
+      msg[DDL_LOG_BUFFER_SIZE - 1] = '\0';
+      retval = DDL_LOG_BUFFER_SIZE;
+    }
+  return retval;
+}
+
+static void
+cub_ddl_log_backup (const char *path)
+{
+  char backup_file[PATH_MAX] = { 0 };
+#if !defined(WINDOWS)
+  sigset_t new_mask, old_mask;
+#endif /* !WINDOWS */
+
+  snprintf (backup_file, sizeof (backup_file), "%s.bak", path);
+
+#if !defined(WINDOWS)
+  sigfillset (&new_mask);
+  sigdelset (&new_mask, SIGINT);
+  sigdelset (&new_mask, SIGQUIT);
+  sigdelset (&new_mask, SIGTERM);
+  sigdelset (&new_mask, SIGHUP);
+  sigdelset (&new_mask, SIGABRT);
+  if (sigprocmask (SIG_SETMASK, &new_mask, &old_mask) == 0)
+    {
+      unlink (backup_file);
+      rename (path, backup_file);
+      sigprocmask (SIG_SETMASK, &old_mask, NULL);
+    }
+#else /* !WINDOWS */
+  unlink (backup_file);
+  rename (path, backup_file);
+#endif
+}
+
+static FILE *
+cub_fopen_and_lock (const char *path, const char *mode)
+{
+#define MAX_RETRY_COUNT 100
+  int retry_count = 0;
+  FILE *ddl_log_fd = NULL;
+
+retry:
+  ddl_log_fd = fopen (path, mode);
+  if (ddl_log_fd != NULL)
+    {
+      if (lockf (fileno (ddl_log_fd), F_TLOCK, 0) < 0)
+	{
+	  fclose (ddl_log_fd);
+	  if (retry_count < MAX_RETRY_COUNT)
+	    {
+	      SLEEP_MILISEC (0, 10);
+	      retry_count++;
+	      goto retry;
+	    }
+	  ddl_log_fd = NULL;
+	}
+    }
+
+  return ddl_log_fd;
+}
+
+#if defined(WINDOWS)
+static void
+unix_style_path (char *path)
+{
+  char *p = NULL;
+  for (p = path; *p; p++)
+    {
+      if (*p == '\\')
+	*p = '/';
+    }
+}
+#endif /* WINDOWS */
+
+static int
+cub_create_dir_log (const char *new_dir)
+{
+  char *p, path[PATH_MAX] = { 0 };
+
+  if (new_dir == NULL)
+    return -1;
+
+  strcpy (path, new_dir);
+  trim (path);
+
+
+#if defined(WINDOWS)
+  unix_style_path (path);
+#endif /* WINDOWS */
+
+  p = path;
+#if defined(WINDOWS)
+  if (path[0] == '/')
+    p = path + 1;
+  else if (strlen (path) > 3 && path[2] == '/')
+    p = path + 3;
+#else /* WINDOWS */
+  if (path[0] == '/')
+    {
+      p = path + 1;
+    }
+#endif /* WINDOWS */
+
+  while (p != NULL)
+    {
+      p = strchr (p, '/');
+      if (p == NULL)
+	return 0;
+
+      if (p != NULL)
+	*p = '\0';
+
+      if (access (path, F_OK) < 0)
+	{
+	  if (mkdir (path, 0777) < 0)
+	    {
+	      return -1;
+	    }
+	}
+      if (p != NULL)
+	{
+	  *p = '/';
+	  p++;
+	}
+    }
+  return 0;
+}
+
+static int
+cub_get_current_time (char *buf, size_t size)
```

Comment:
What about

`cub_ddl_log_get_current_date_time_string (char *buf, size_t size)`

Reply:
Thank you, i will change code as your recommend.